presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/Rule.java

```diff
@@ -14,24 +14,32 @@
 package com.facebook.presto.sql.planner.iterative;
 
 import com.facebook.presto.Session;
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Match;
 import com.facebook.presto.matching.Pattern;
 import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
 import com.facebook.presto.sql.planner.SymbolAllocator;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 
 import java.util.Optional;
 
-public interface Rule
+public interface Rule<T>
 {
     /**
      * Returns a pattern to which plan nodes this rule applies.
      */
-    default Pattern getPattern()
+    Pattern<T> getPattern();
+
+    default Optional<PlanNode> apply(PlanNode node, Context context)
```

Comment:
IMO iterative optimizer should call directly the second  `apply` method so this one is not needed. I see no point of calling `match` twice per matching rule